README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/qsort)](https://cran.r-project.org/package=qsort) [![Rdoc](http://www.rdocumentation.org/badges/version/qsort)](http://www.rdocumentation.org/packages/qsort) [![Build Status](https://travis-ci.org/joaordaniel/qsort.svg?branch=master)](https://travis-ci.org/joaordaniel/qsort) [![DOI](https://zenodo.org/badge/132877059.svg)](https://zenodo.org/badge/latestdoi/132877059)

qsort
=====

Overview
--------

`qsort` is a package that allows scoring Q-sort data, using criteria sorts and derived scales from subsets of items. This package includes descriptions and scoring procedures for four different Q-sets:

-   Attachment Q-set (version 3.0) (Waters, 1995);
-   California Child Q-set (Block & Block, 1969);
-   Maternal Behaviour Q-set (version 3.1) (Pederson et al., 1999);
-   Preschool Q-set (Baumrind, 1968 revised by Wanda Bronson).

qsort package includes 4 objects:

-   `qsort_score()` a function for scoring Q-sort data;
-   `qsets` a list with four data frames used in `qsort_score()`, containing criterion scores and scales for the referred Q-sets;
-   `ex_qsort` a list containing four example data frames for the referred Q-sets;
-   `print_cards()` a function for printing Q-set item cards.

Read `qsets` help file (`?qsets`) to learn more about which criteria sorts and scales are included for each Q-set.

Installation
------------

The `qsort` package can be installed from GitHub:

``` r
# to install packages from github you first need to install devtools package from CRAN.
# if you haven't installed devtools already just type:
install.packages("devtools")

# to install qsort from github type:
devtools::install_github("joaordaniel/qsort")
```

To install `qsort` package from CRAN:

``` r
install.packages("qsort")
```

Example
-------

The example bellow shows how to use `qsort_score()` function to compute scores from California Child Q-sort data, present in `ex_qsort` datasets (`ex_qsort$ccq`).

``` r
library(qsort)
data_ccq <- qsort_score(ex_qsort$ccq, qset = "ccq", item1 = "ccq1", subj_id = "participant", group_id = "classroom")
data_ccq
```

    ##    qset participant classroom scomp_c sest_c egores_c egocont_c sdes_c
    ## 1   ccq           1         1  -0.074 -0.093   -0.145    -0.012 -0.137
    ## 2   ccq           2         1  -0.023  0.008    0.090     0.157  0.053
    ## 3   ccq           3         1   0.092  0.086    0.112    -0.021  0.132
    ## 4   ccq           4         1  -0.105 -0.113   -0.182    -0.128 -0.160
    ## 5   ccq           5         1  -0.010 -0.039   -0.092    -0.092 -0.053
    ## 6   ccq           6         2  -0.104 -0.079   -0.042     0.156 -0.089
    ## 7   ccq           7         2   0.051  0.079    0.168     0.169  0.124
    ## 8   ccq           8         2   0.049  0.066    0.153     0.185  0.118
    ## 9   ccq           9         2  -0.024 -0.007    0.009    -0.026  0.007
    ## 10  ccq          10         2  -0.039 -0.046   -0.033     0.098 -0.042
    ##    partial_scomp_c partial_sest_c partial_egores_c partial_egocont_c
    ## 1            0.113          0.047           -0.054            -0.007
    ## 2           -0.160         -0.073            0.086             0.155
    ## 3           -0.061         -0.053           -0.002            -0.026
    ## 4            0.088          0.047           -0.089            -0.123
    ## 5            0.085          0.012           -0.090            -0.090
    ## 6           -0.054         -0.004            0.068             0.160
    ## 7           -0.139         -0.053            0.121             0.166
    ## 8           -0.130         -0.068            0.101             0.182
    ## 9           -0.069         -0.026            0.004            -0.026
    ## 10          -0.003         -0.019            0.005             0.100
    ##    shields_s_emreg
    ## 1              4.3
    ## 2              5.9
    ## 3              5.1
    ## 4              3.7
    ## 5              4.4
    ## 6              5.7
    ## 7              5.9
    ## 8              5.9
    ## 9              5.0
    ## 10             5.2

Read `qsort_score` help file (`?qsort_score`) to learn more about the function's four arguments, and `qsets` help file (`?qsets`) to learn more about variables' names.

Print cards
-----------

To print item descriptions in separate cards use `print_cards()` function. The example bellow uses the `print_cards()` function to create a pdf file in a user defined directory (e.g., "C:/Users/jd/R/") with Attachment Q-set items.

``` r
library(qsort)
print_cards(qset = "aqs", desc_col = "description", dir.print = "C:/Users/jd/R/")
```

Read `print_cards()` help file (`?print_cards`) to learn more about the function's two arguments.

References
----------

Baumrind, D. (1968). Manual for the Preschool Behaviour Q-set. Parental Research Project. Berkeley, CA: Institute of Human Development.

Block, J. H., & Block, J. (1969). The California Child Q-Set. Berkeley, CA: Institute of Human Development, University of California.

Pederson, D. R., Moran, G., & Bento, S. (1999). Maternal Behaviour Q-sort (version 3.1). London, ON: Psychology Department, Western University.

Waters, E. (1995). Appendix A: The attachment Q-set (Version 3.0). Monographs of the Society for Research in Child Development, 60, 234-246.


NEWS.md
<!-- NEWS.md is generated from NEWS.Rmd. Please edit that file -->
qsort 0.2.2 (30/07/2018)
========================

Minor Changes
-------------

-   Added Shields and Cicchetti’s (1997) emotion regulation scale to
    qsets$ccq

qsort 0.2.1 (11/06/2018)
========================

Major Changes
-------------

-   Added `print_cards()` function for printing Q-set item cards.

qsort 0.2.0 (24/05/2018)
========================

Major Changes
-------------

-   Added Attachment Q-set (version 3.0) (Waters, 1995) and  
    Maternal Behaviour Q-set (version 3.1) (Pederson et al., 1999) to
    `qsets`.
-   Added ego-resiliency and ego-control criterion sorts to `qsets$ccq`
-   `qsort_score()` now computes scales scores as well. The source code
    of this function was substantially re-written.

qsort 0.1.0
===========

First release


R/qsets.R
#'Q-sets criteria scores and derived scales
#'
#'A list containing four different datasets: Attachment Q-set (aqs; version
#'3.0), California Child Q-set (ccq); Maternal Behaviour Q-set (mbqs; version
#'3.1); and Preschool Q-set (pq). In these datasets, column names ending in "_c"
#'refer to criteria sorts' scores, while column names ending in "_s" refer to
#'scales' scores.
#'
#'@format
#' \enumerate{
#'  \item \strong{qsets$aqs} AQS dataset with 90 rows (items)
#'  and 9 columns.
#'  \describe{
#'    \item{item}{Item number from 1 to 90.}
#'    \item{qset}{aqs. Attachment Q-set (version 3.0) (Waters, 1995).}
#'    \item{sec_c}{Security criterion scores
#'    (retrieved from Waters et al., 1995).}
#'    \item{dep_c}{Dependency criterion scores
#'    (retrieved from Waters et al., 1995).}
#'    \item{pederson_s}{Pederson & Moran's (1995) scales
#'    (retrieved from Waters et al., 1995). comp = compliance; sb = secure base,
#'    fd = fussy/difficult; epc = enjoys physical contact; as = affective sharing.}
#'    \item{pederson_s_inv}{Items to be inverted in Pederson and Moran scales.}
#'    \item{posada_s}{Posada et al.'s (1995) scales (retrieved from Waters et al.,
#'    1995). sim = smooth interactions with mother; pcm = physical contact with
#'    mother; ioa = interactions with other adults; pm = proximity to mother.}
#'    \item{posada_s_inv}{Items to be inverted in Posada et al.'s scales.}
#'    \item{description}{Item description (retrieved from Waters, 1995).} }
#'
#'  \item \strong{qsets$ccq} CCQ dataset with 100 rows (items) and 6 columns.
#'  \describe{
#'    \item{item}{Item number from 1 to 100.}
#'    \item{qset}{ccq. California Child Q-set (Block & Block, 1969).}
#'    \item{scomp_c}{Social competence criterion scores
#'    (retrieved from Waters et al., 1985).}
#'    \item{sest_c}{Self-esteem criterion scores
#'    (retrieved from Waters et al., 1985).}
#'    \item{sest_c}{Self-esteem criterion scores
#'    (retrieved from Waters et al., 1985).}
#'    \item{egores_c}{Ego-resiliency criterion scores (Block & Block, 1980;
#'    retrieved from Brian Vaughn, personal communication).}
#'    \item{egocont_c}{Ego-control criterion scores (Block & Block, 1980;
#'    retrieved from Brian Vaughn, personal communication).}
#'    \item{shields_s}{Emotion regulation scale
#'    (retrieved from Shields and Cicchetti, 1997)}
#'    \item{shields_s_inv}{Items to be inverted in emotion regulation scale.}
#'    \item{description}{Item description (retrieved from Waters et al., 1985).} }
#'
#'  \item \strong{qsets$mbqs} MBQS dataset with 90 rows (items) and 4 columns.
#'  \describe{
#'    \item{item}{Item number from 1 to 90.}
#'    \item{qset}{mbqs. Maternal Behaviour Q-set (version 3.1) (Pederson et al., 1999).}
#'    \item{sens_s}{Sensibility criterion scores
#'    (retrieved from Pederson et al., 1999).}
#'    \item{description}{Item description
#'    (retrieved from Pederson et al., 1999).} }
#'
#'  \item \strong{qsets$pq} PQ dataset with 72 rows (items) and 6 columns.
#'  \describe{
#'    \item{item}{Item number from 1 to 72.}
#'    \item{qset}{pq. Preschool
#'  Q-set (Baumrind, 1968 revised by Wanda Bronson).}
#'    \item{scomp_c}{Social competence criterion scores
#'    (retrieved from Waters et al., 1985).}
#'    \item{sest_c}{Self-esteem criterion scores
#'    (retrieved from Waters et al., 1985).}
#'    \item{sdes_c}{Social desirability criterion scores
#'    (retrieved from Waters et al., 1985).}
#'    \item{description}{Item description (retrieved from Waters et al., 1985).} } }
#'
#'@references Baumrind, D. (1968). Manual for the Preschool Behaviour Q-set.
#'  Parental Research Project. Berkeley, CA: Institute of Human Development,
#'  University of California.
#'
#'  Block, J. H., & Block, J. (1969). The California Child Q-Set. Berkeley, CA:
#'  Institute of Human Development, University of California.
#'
#'  Block, J. H., & Block, J. (1980). The role of ego-control and ego-resiliency
#'  in the organization of behavior. In W. A. Collins (Ed.), Minnesota symposium
#'  on child psychology: Vol. 13. Development of cognition affect, and social
#'  relations (pp. 39–101), Hillsdale, NJ: Erlbaum.
#'
#'  Posada, G., Waters, E., Crowell, J. A., & Lay, K. (1995). Is it easier to
#'  use a secure mother as a secure base? Attachment Q-Sort correlates of the
#'  adult attachment interview. Monographs of the Society for Research in Child
#'  Development, 60 (2/3), 133-145.
#'
#'  Pederson, D. R., & Moran, G. (1995). A categorical description of
#'  infant-mother relationships in the home and its relation to Q-sort measures
#'  of infant-mother interaction. Monographs of the Society for Research in
#'  Child Development, 60 (2/3), 111-132.
#'
#'  Pederson, D. R., Moran, G., & Bento, S. (1999). Maternal Behaviour Q-sort
#'  (version 3.1). London, ON: Psychology Department, Western University.
#'
#'  Shields, A., & Cicchetti, D. (1997). Emotion regulation among school-age
#'  children: The development and validation of a new criterion q-sort scale.
#'  Developmental Psychology, 33, 906-916.
#'
#'  Waters, E. (1995). Appendix A: The attachment Q-set (Version 3. 0).
#'  Monographs of the Society for Research in Child Development, 60 (2/3),
#'  234-246.
#'
#'  Waters, E., Vaughn, B. E., Posada, G., & Kondo-Ikemura, K. (1995). Appendix
#'  D: Scoring key for Q-sort criteria and derived scales in the attachment
#'  Q-set and the attachment Q-set for infant macaques item sets. Monographs of
#'  the Society for Research in Child Development, 60 (2/3), 280-282.
#'
#'  Waters, E., Noyes, D. M., Vaughn, B. E., & Ricks, M. (1985). Q-sort
#'  definitions of social competence and self-esteem: Discriminant validity of
#'  related constructs in theory and data. Developmental Psychology, 21,
#'  508-522.
#'
"qsets"


R/ex_qsort.R
#'Example datasets for qsort package
#'
#'A list containing four hypothetical Q-sort datasets:
#'Attachment Q-set (aqs; version 3.0) (Waters, 1995);
#'California Child Q-set (ccq; Block and Block, 1969);
#'Maternal Behaviour Q-set (version 3.1) (mbqs; Pederson et al., 1999);
#'Preschool Q-set (pq; Baumrind, 1968 revised by Wanda Bronson).
#'
#'For these examples, items were given random scores (1 to 9; least to most
#'characteristic respectively) following a rectangular distribution. For the ccq
#'Q-set this corresponds to placing 11 items in each of the 9 categories, plus
#'an additional one in the center category (i.e., 12 items in category 5). For
#'the remaining Q-sets all the 9 categories include the same number of items
#'(aqs = 10, mbqs = 10, pq = 8).
#'
#' @format Data frames with 10 rows and a variable number of columns, depending
#'   on the number of Q-set items (aqs = 90, ccq = 100, mbqs = 90, pq = 72)
#'   \describe{
#'   \item{participant}{Subject identification.}
#'   \item{classroom}{Group identification.}
#'   \item{qset1 to ...}{Items' scores.}
#'   }
#'
#'@references Baumrind, D. (1968). Manual for the Preschool Behaviour Q-set.
#'  Parental Research Project. Berkeley, CA: Institute of Human Development,
#'  University of California.
#'
#'  Block, J. H., & Block, J. (1969). The California Child Q-Set. Berkeley, CA:
#'  Institute of Human Development, University of California.
#'
#'  Pederson, D. R., Moran, G., & and Bento, S. (1999). Maternal Behaviour
#'  Q-sort (version 3.1). London, ON: Psychology Department, Western University.
#'
#'  Waters, E. (1995). Appendix A: The attachment Q-set (Version 3. 0).
#'  Monographs of the Society for Research in Child Development, 60, 234-246.
"ex_qsort"





R/criterion.R
#' Waters et al. (1985) criterion scores
#'
#' A dataset containing Waters et al.'s (1985) social competence, self-esteem
#' and social desirability criterion scores for the California Child Q-set (CCQ;
#' Block and Block, 1969) and the Preschool Q-set (PQ; Baumrind, 1968 revised by
#' Wanda Bronson). Q-set items are scored 1 to 9 (most uncharacteristic to most
#' characteristic of the respective domain). Originally Waters et al. (1985)
#' scored the most characteristic items as 1 and the least characteristic ones
#' as 9. Here we inverted the scores to be more intuitive. qsort_score function
#' uses this dataset to compute social competence and self-esteem scores.
#' Consequently, changes made in this dataset will affect qsort_score output. We
#' made a few corrections to Water et al.'s (1985) items' numbers and
#' descriptions: (1) CCQ item numbers 15 and 96 appear repeated twice - "Shows
#' concerns for moral issues (reciprocity, fairness)" is now CCQ item number 25,
#' and "Is verbally fluent" is now CCQ item 69; (2) PQ item 23 now reads "Other
#' children seek his company", instead of "Other children seldom seek his
#' company".
#'
#' @format A data frame with 172 rows and 6 variables: \describe{
#'   \item{item}{item number, ranging from 1 to 100 for CCQ items and 1 to 72
#'   for PQ items} \item{qset}{ccq - California Child Q-set; pq - Preschool
#'   Q-set} \item{scomp_c}{social competence criterion scores}
#'   \item{sest_c}{self-esteem competence criterion scores} \item{sdes_c}{social
#'   desirability criterion scores} \item{description}{item description} }
#'
#' @references Baumrind, D. (1968). Manual for the Preschool Behaviour Q-set.
#'   Parental Research Project. Berkeley, CA: Institute of Human Development,
#'   University of California.
#'
#'   Block, J. H., & Block, J. (1969). The California Child Q-Set. Berkeley, CA:
#'   Institute of Human Development, University of California.
#'
#'   Waters, E., Noyes, D. M., Vaughn, B. E., & Ricks, M. (1985). Q-sort
#'   definitions of social competence and self-esteem: Discriminant validity of
#'   related constructs in theory and data. Developmental Psychology, 21,
#'   508-522.
#'
"criterion"


R/print_cards.R
#' print_cards
#'
#' print_cards creates a pdf document in a specified directory, with item’s
#' descriptions displayed on separate cards.
#'
#' @param qset "aqs" for Attachment Q-set (version 3.0) (Waters, 1995); "ccq"
#'   for California Child Q-set (Block & Block, 1969); "mbqs" for Maternal
#'   Behaviour Q-set (version 3.1) (Pederson et al., 1999); and "pq" for
#'   Preschool Q-set (Baumrind, 1968 revised by Wanda Bronson).
#' @param desc_col Column name of qsets data frame containing item's
#'   descriptions.
#'
#' @param dir.print Directory path where .pdf file will be saved
#'
#' @return a pdf document with item’s descriptions displayed on separate cards.
#'
#' @export
#'
#' @importFrom ggplot2 element_blank ggplot theme_bw theme scale_x_continuous
#'   scale_y_continuous coord_fixed annotate
#'
#' @importFrom gridExtra marrangeGrob
#'
#' @importFrom cowplot save_plot

#' @examples
#' \donttest{print_cards(qset = "aqs", desc_col = "description", dir.print = tempdir())}
#'
#'@references Baumrind, D. (1968). Manual for the Preschool Behaviour Q-set.
#'  Parental Research Project. Berkeley, CA: Institute of Human Development,
#'  University of California.
#'
#'  Block, J. H., & Block, J. (1969). The California Child Q-Set. Berkeley, CA:
#'  Institute of Human Development, University of California.
#'
#'  Pederson, D. R., Moran, G., & Bento, S. (1999). Maternal Behaviour Q-sort
#'  (version 3.1). London, ON: Psychology Department, Western University.
#'
#'  Waters, E. (1995). Appendix A: The attachment Q-set (Version 3. 0).
#'  Monographs of the Society for Research in Child Development, 60, 234-246.

print_cards <- function(qset, desc_col = "description", dir.print){
cards <- list()
for(i in 1:nrow(qsort::qsets[[qset]])){
# add item number to item description
    item_desc <- paste(i,
                     qsort::qsets[[qset]][[desc_col]][[i]],
                     sep=". ")
# break item description in 40 characters' sentences
  item_desc_wrap <- strwrap(item_desc, width = 40)
# define y axis position of each sentence
# y axis position decreases 0.4 for each additional sentence
  desc_ypos <- seq(2, length.out = length(item_desc_wrap), by = -0.4)

# create an empty plot for each item description
  cards[[i]] <- ggplot2::ggplot() +
    ggplot2::theme_bw() +
    ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
          panel.grid.minor = ggplot2::element_blank(),
          axis.ticks = ggplot2::element_blank(),
          axis.text = ggplot2::element_blank(),
          axis.title = ggplot2::element_blank()) +
    ggplot2::scale_x_continuous(limits = c(0, 3.5)) +
    ggplot2::scale_y_continuous(limits = c(0, 2.5)) +
    ggplot2::coord_fixed(ratio = 1) +
# paste item description (item_desc_wrap)
    ggplot2::annotate("text", x = 1.75, y = desc_ypos, size = 4, label = item_desc_wrap)
}

# arrange 9 cards/plots per page
cards2 <- gridExtra::marrangeGrob(cards, ncol = 3, nrow = 3)
# save plots in a pdf file in the working directory
cowplot::save_plot(filename = paste(dir.print, "cards_", qset, ".pdf", sep = ""),
                   cards2, ncol = 3, nrow = 3,
                   base_height = 2.5, base_width = 3.5)
}



R/qsort_score.R
#'qsort_score
#'
#'qsort_score returns a data frame with scores derived from criteria Q-sorts and
#'from scales created from subsets of items.
#'
#'@param x data frame. x should be in wide format, with each subject data in one
#'  row and the scores for qset items in different columns. Columns for qset
#'  items' scores should be ordered sequentially from left to right, starting
#'  with item 1.
#'
#'@param qset "aqs" for Attachment Q-set (version 3.0) (Waters, 1995); "ccq" for
#'  California Child Q-set (Block & Block, 1969); "mbqs" for Maternal
#'  Behaviour Q-set (version 3.1) (Pederson et al., 1999); and "pq" for
#'  Preschool Q-set (Baumrind, 1968 revised by Wanda Bronson).
#'@param item1 Column name of x containing item 1 score
#'@param subj_id Optional. Column name of x with subjects' identification codes.
#'@param group_id Optional. Column name of x with groups' identification codes.
#'
#'@return qsort_score function returns a data frame. This data frame will have a
#'  varying number of columns depending on the number of available criteria
#'  sorts and scales for the selected Q-set. Column names ending in "_c" refer
#'  to criteria sorts' scores, while column names ending in "_s" refer to
#'  scales' scores. Following Waters et al.'s (1985) suggestion about the
#'  influence of social desirability bias in Q-sort data, for Q-sets that have
#'  social desirability criterion scores, this data frame will also include
#'  criteria sorts' scores controlled for social desirability (i.e., partial
#'  correlations). Column names starting with "partial_" refer to these scores.
#'  The different criteria sorts and scales for each Q-set can be consulted in
#'  the documentation of qsets object (?qsets).
#'
#'@export
#'
#'@importFrom purrr modify_if
#'@importFrom stats aggregate.data.frame cor
#'
#'@examples
#'data_ccq <- qsort_score(ex_qsort$ccq,
#'                        qset = "ccq",
#'                        item1 = "ccq1",
#'                        subj_id = "participant",
#'                        group_id = "classroom")
#'data_ccq
#'
#'data_aqs <- qsort_score(ex_qsort$aqs,
#'                        qset = "aqs",
#'                        item1 = "aqs1")
#'
#'@references Baumrind, D. (1968). Manual for the Preschool Behaviour Q-set.
#'  Parental Research Project. Berkeley, CA: Institute of Human Development,
#'  University of California.
#'
#'  Block, J. H., & Block, J. (1969). The California Child Q-Set. Berkeley, CA:
#'  Institute of Human Development, University of California.
#'
#'  Pederson, D. R., Moran, G., & Bento, S. (1999). Maternal Behaviour
#'  Q-sort (version 3.1). London, ON: Psychology Department, Western University.
#'
#'  Waters, E. (1995). Appendix A: The attachment Q-set (Version 3. 0).
#'  Monographs of the Society for Research in Child Development, 60, 234-246.
#'
#'  Waters, E., Noyes, D. M., Vaughn, B. E., & Ricks, M. (1985). Q-sort
#'  definitions of social competence and self-esteem: Discriminant validity of
#'  related constructs in theory and data. Developmental Psychology, 21, 508-522.

qsort_score <- function(x, qset = names(qsort::qsets), item1, subj_id = NULL, group_id = NULL) {
  # store number of items of each qset
  qsort_length <- c(90, 100, 72, 90)
  names(qsort_length) <- c("aqs", "ccq", "pq", "mbqs")

# assures that x is a data frame
# it will turn a tibble back to a data.frame for instance
  x <- as.data.frame(x)

# create a selection vector with column numbers of x qset items
  start_x <- match(item1, (colnames(x)))
  end_x <- start_x + qsort_length[[qset]] - 1
  sel_x <- start_x : end_x

# identify criteria scores' column(s) in selected qset
# column names ending in "_c"
  cscores <- grep("_c$", names(qsort::qsets[[qset]]), value = T)

# identify scales' column(s) in selected qset
# column names ending in "_s"
  scales <- grep("_s$", names(qsort::qsets[[qset]]), value = T)

# identify column(s) detailing items to be inverted for scale computation
# column names ending in "_inv"
  scales_inv <- grep("_inv$", names(qsort::qsets[[qset]]), value = T)

# create an empty data frame to store output values
  qsort_data <- data.frame()

  for(i in 1:nrow(x)){
# for all rows/cases in x
# first create empty data frames to store temporary values used in the computations below
    temp_id <- cbind.data.frame(qset, x[i, subj_id], x[i, group_id])
    names(temp_id) <- c("qset", subj_id, group_id)
    temp_cs <- data.frame()[1, ]
    temp_s <- data.frame()[1,]
    temp_s3 <- data.frame()[1, ]

    if(length(cscores) > 0){
# if selected qset contains criteria scores
      for(j in 1:length(cscores)){
# compute the correlation between the individual profile and each criterion sort
# store correlation values in j number of columns in temp_cs
# and name them accordingly
        temp_cs[1, j] <- cor(t(x[i, sel_x]),
                         qsort::qsets[[qset]][ , cscores[[j]]],
                         use = "pairwise.complete.obs")
      }
      names(temp_cs) <- cscores

      if("sdes_c" %in% cscores){
# if one of the criteria sorts corresponds to social desirability
        cscores2 <- cscores[cscores != "sdes_c"]
        for(k in 1:length(cscores2)){
# compute partial correlations between the individual profile data and each criterion sort,
# while controlling for social desirability
# add partial correlation values to temp_cs
# and name columns accordingly
          cor_ab <- temp_cs[cscores2[[k]]]
          cor_ac <- temp_cs["sdes_c"]
          cor_bc <- stats::cor(qsort::qsets[[qset]][ , cscores2[[k]]],
                           qsort::qsets[[qset]][ , "sdes_c"],
                           use = "pairwise.complete.obs")

          temp_cs[1, (length(cscores) + k)] <- (cor_ab - cor_ac * cor_bc) /
                                                (sqrt(1 - cor_ac^2) * sqrt(1 - cor_bc^2))
        }
        names(temp_cs) <- c(cscores, paste("partial_", cscores2, sep =""))
      }
    }

    if(length(scales) > 0){
# if selected qset contains scales
      for(j in 1:length(scales)){
# match individual profile data with columns indicating scales' items
# and which items need to be inverted
        temp_s <- cbind.data.frame(as.numeric(t(x[i, sel_x])),
                        qsort::qsets[[qset]][ , scales[j]],
                        qsort::qsets[[qset]][ , scales_inv[j]])
        names(temp_s) <- c("item_score", "scales", "scales_inv")

# invert items and compute scales' scores
        temp_s$item_score <- ifelse(temp_s$scales_inv == 1,
                                    10 - temp_s$item_score,
                                    temp_s$item_score)

        temp_s <- stats::aggregate.data.frame(temp_s$item_score,
                                       by = list(scales = temp_s$scales),
                                       FUN = mean, na.rm = T)

# store each score in a separate column of temp_s2
# name them accordingly
        temp_s2 <- as.data.frame(t(temp_s[[2]]))
        names(temp_s2) <- paste(scales[j], "_", temp_s[[1]], sep = "")

# bind scores for all scales presented in selected qset
        temp_s3 <- cbind.data.frame(temp_s3, temp_s2)
      }
    }

# bind each case identification columns (if given),
# correlation values computed from criteria sorts,
# and scale scores
    temp_data <- cbind.data.frame(temp_id, temp_cs, temp_s3)

# add a row to output data frame
    qsort_data <- rbind(qsort_data, temp_data)
  }

  row.names(qsort_data) <- NULL

# round numeric variables to three digits
  qsort_data <- purrr::modify_if(qsort_data, ~is.numeric(.), ~round(., 3))

# remove attributes of numeric variables
  for(i in 1:ncol(qsort_data)){
    if(is.numeric(qsort_data[[i]])){
      attributes(qsort_data[[i]]) <- NULL
    }
  }

# return output
  qsort_data
}
